File: 00_Alternate_Languages/96_Word/perl/word.pl
#!/usr/bin/env perl

use 5.010;      # To get 'state' and 'say'

use strict;     # Require explicit declaration of variables
use warnings;   # Enable optional compiler warnings

use English;    # Use more friendly names for Perl's magic variables
use Term::ReadLine;     # Prompt and return user input

our $VERSION = '0.000_01';

print <<'EOD';
                                 WORD
               Creative Computing  Morristown, New Jersey



I am thinking of a word -- you guess it.  I will give you
clues to help you get it.  Good luck!!


EOD

# Read the content of __DATA__, remove the trailing newlines, and store
# each line into @words. Stop at __END__, since Perl does not see this
# as an end-of-file.
my @words;
while ( <DATA> ) {
    chomp;
    last if $ARG eq '__END__';
    push @words, lc $ARG;   # Normalize case to lower.
}

# This loop represents an actual game. We execute it until the player
# does something that makes us explicitly break out.
while ( 1 ) {
    print <<'EOD';


You are starting a new game ...
EOD

    # Choose a random target word. The rand() function returns a number
    # from 0 to its argument, and coerces its argument to a scalar. In
    # scalar context, an array evaluates to the number of elements it
    # contains.
    my $target = $words[ rand @words ];

    # We generalize the code by using the actual length of the target.
    my $target_length = length $target;

    my $count = 0;      # Number of guesses

    # Make an array of the individual letters in the target. We will
    # iterate over this to determine matching letters.
    my @target_array = split qr<>, $target;

    # Make a hash of those letters. We will use this to determine common
    # letters. Any true value will do for the value of the hash. By
    # making use of this hash we avoid the nested loops of the original
    # BASIC program.
    my %target_hash = map { $ARG => 1 } @target_array;

    # We keep prompting the player until we get a response that causes
    # us to break out of the loop.
    while ( 1 ) {

        # Create the readline object. The state keyword means the
        # variable is only initialized once, no matter how many times
        # execution passes this point.
        state $term = Term::ReadLine->new( 'word' );

        # Read the next guess. A return of undef means end-of-file.
        my $guess = $term->readline( "Guess a $target_length letter word: " );
        exit unless defined $guess;

        last if $guess eq '?';  # A question mark means we give up
        if ( length( $guess ) != $target_length ) {
            # Wrong length. Ask again.
            say "You must guess a $target_length letter word.  Try again.";
            redo;       # Redo the innermost loop
        }

        $guess = lc $guess;     # Lower-case the guess
        $count++;       # Count another guess

        if ( $guess eq $target ) {
            # We guessed the word.
            say "You have guessed the word. It took $count guesses!";
            my $answer = $term->readline( 'Want to play again? [y/N]: ');
            exit unless defined $guess; # End of file
            exit unless $guess =~ m/ \A y /smxi;
            last;       # Exit the innermost loop.
        }

        my @common_letters; # Letters common to guess and target
        my $match = '-' x length $target;   # Assume no matches
        my $inx = 0;    # Iterator
        foreach my $letter ( split qr<>, $guess ) {
            if ( $target_hash{$letter} ) {
                # If the letter is in the hash, it occurs in the target
                push @common_letters, $letter;
                # If it is at the current position in the target, it is
                # an actual match.
                $target_array[$inx] eq $letter
                    and substr $match, $inx, 1, $letter;
            }
            $inx++;
        }

        say 'There were ', scalar @common_letters,
            ' matches and the common letters were... ', @common_letters;
        say "From the exact letter matches, you know................ $match";
        say '';
        say q<If you give up, type '?' for your next guess.>;
        redo;
    }

}
__DATA__
dinky
smoke
water
grass
train
might
first
candy
champ
would
clump
dopey
__END__

=head1 TITLE

word.pl - Play the game 'word' from Basic Computer Games

=head1 SYNOPSIS

 word.pl

=head1 DETAILS

This Perl script is a port of C<word>, which is the 96th entry in Basic
Computer Games.

=head1 PORTED BY

Thomas R. Wyant, III F<wyant at cpan dot org>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2022 by Thomas R. Wyant, III

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl 5.10.0. For more details, see the Artistic
License 1.0 at
L<https://www.perlfoundation.org/artistic-license-10.html>, and/or the
Gnu GPL at L<http://www.gnu.org/licenses/old-licenses/gpl-1.0.txt>.

This program is distributed in the hope that it will be useful, but
without any warranty; without even the implied warranty of
merchantability or fitness for a particular purpose.

=cut

# ex: set expandtab tabstop=4 textwidth=72 :

File: 00_Alternate_Languages/96_Word/javascript/word.js
// WORD
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{

    return new Promise(function (resolve) {
                       const input_element = document.createElement("INPUT");

                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode === 13) {
                                                          const input_str = input_element.value;
                                                          document.getElementById("output").removeChild(input_element);
                                                          print(input_str);
                                                          print("\n");
                                                          resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    let str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// These are the words that the game knows about> If you want a bigger challenge you could add more words to the array
const WORDS = ["DINKY", "SMOKE", "WATER", "GLASS", "TRAIN",
             "MIGHT", "FIRST", "CANDY", "CHAMP", "WOULD",
             "CLUMP", "DOPEY"];
const WORD_COUNT = WORDS.length;

// Main control section
async function main()
{
    print(tab(33) + "WORD\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("I AM THINKING OF A WORD -- YOU GUESS IT.  I WILL GIVE YOU\n");
    print("CLUES TO HELP YOU GET IT.  GOOD LUCK!!\n");
    print("\n");
    print("\n");
    outer: while (1) {
        print("\n");
        print("\n");
        print("YOU ARE STARTING A NEW GAME...\n");

        const secretWord = WORDS[Math.floor(Math.random() * WORD_COUNT)];

        let guessCount = 0;
        // This array holds the letters which have been found in the correct position across all guesses
        // For instance if the word is "PLAIN" and the guesses so far are
        // "SHALL" ("A" correct) and "CLIMB" ("L" correct) then it will hold "-LA--"
        const knownLetters = [];
        for (let i = 0; i < 5; i++)
            knownLetters[i] = "-";

        let guess = undefined;
        while (1) {
            print("GUESS A FIVE LETTER WORD");
            guess = (await input()).toUpperCase();
            guessCount++;
            if (secretWord === guess) {
                // The player has guessed correctly
                break;
            }

            if (guess.charAt(0) === "?") {
                // Player has given up
                print("THE SECRET WORD IS " + secretWord + "\n");
                print("\n");
                // Start a new game by going to the start of the outer while loop
                continue outer;
            }

            if (guess.length !== 5) {
                print("YOU MUST GUESS A 5 LETTER WORD.  START AGAIN.\n");
                print("\n");
                guessCount--;
                continue;
            }

            // Two things happen in this double loop:
            // 1. Letters which are in both the guessed and secret words are put in the lettersInCommon array
            // 2. Letters which are in the correct position in the guessed word are added to the knownLetters array
            let lettersInCommonCount = 0;
            const lettersInCommon = [];
            for (let i = 0; i < 5; i++) {// loop round characters in secret word
                let secretWordCharacter = secretWord.charAt(i);
                for (let j = 0; j < 5; j++) {// loop round characters in guessed word
                    let guessedWordCharacter = guess.charAt(j);
                    if (secretWordCharacter === guessedWordCharacter) {
                        lettersInCommon[lettersInCommonCount] = guessedWordCharacter;
                        if (i === j) {
                            // Letter is in the exact position so add to the known letters array
                            knownLetters[j] = guessedWordCharacter;
                        }
                        lettersInCommonCount++;
                    }
                }
            }

            const lettersInCommonText = lettersInCommon.join("");
            print("THERE WERE " + lettersInCommonCount + " MATCHES AND THE COMMON LETTERS WERE... " + lettersInCommonText + "\n");

            const knownLettersText = knownLetters.join("");
            print("FROM THE EXACT LETTER MATCHES, YOU KNOW............ " + knownLettersText + "\n");

            if (knownLettersText === secretWord) {
                guess = knownLettersText;
                break;
            }

            if (lettersInCommonCount <= 1) {
                print("\n");
                print("IF YOU GIVE UP, TYPE '?' FOR YOUR NEXT GUESS.\n");
                print("\n");
            }
        }

        print("YOU HAVE GUESSED THE WORD.  IT TOOK " + guessCount + " GUESSES!\n");
        print("\n");

        print("WANT TO PLAY AGAIN");
        const playAgainResponse = (await input()).toUpperCase();
        if (playAgainResponse !== "YES")
            break;
    }
}

main();
